### ISSUE TITLE
Roadrunner breaks with distrib uncertainties

### ISSUE THREAD

**Issue by matthiaskoenig** [sys-bio/roadrunner#707]
Hi all,
when having models with uncertainties in SBML distrib roadrunner breaks.
The uncertainties are only advanced annotations so should be ignored by roadrunner, but not break  model compilation with LLVM.

Minimal example attached:
[simvastatin_liver.zip](https://github.com/sys-bio/roadrunner/files/5473779/simvastatin_liver.zip)
```
import roadrunner

r = roadrunner.RoadRunner("simvastatin_liver_uncertainty.xml")
```
Results in 
```
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double)* @rr_distrib_uniform
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double)* @rr_distrib_normal
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double, double)* @rr_distrib_normal_four
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double)* @rr_distrib_binomial
Global is external, but doesn't have external or weak linkage!
double (i8*, double)* @rr_distrib_bernoulli
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double, double)* @rr_distrib_binomial_four
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double)* @rr_distrib_cauchy
Global is external, but doesn't have external or weak linkage!
double (i8*, double)* @rr_distrib_cauchy_one
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double, double)* @rr_distrib_cauchy_four
Global is external, but doesn't have external or weak linkage!
double (i8*, double)* @rr_distrib_chisquare
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double)* @rr_distrib_chisquare_three
Global is external, but doesn't have external or weak linkage!
double (i8*, double)* @rr_distrib_exponential
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double)* @rr_distrib_exponential_three
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double)* @rr_distrib_gamma
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double, double)* @rr_distrib_gamma_four
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double)* @rr_distrib_laplace
Global is external, but doesn't have external or weak linkage!
double (i8*, double)* @rr_distrib_laplace_one
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double, double)* @rr_distrib_laplace_four
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double)* @rr_distrib_lognormal
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double, double)* @rr_distrib_lognormal_four
Global is external, but doesn't have external or weak linkage!
double (i8*, double)* @rr_distrib_poisson
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double)* @rr_distrib_poisson_three
Global is external, but doesn't have external or weak linkage!
double (i8*, double)* @rr_distrib_rayleigh
Global is external, but doesn't have external or weak linkage!
double (i8*, double, double, double)* @rr_distrib_rayleigh_three
LLVM ERROR: Broken module found, compilation aborted!
```
Best Matthias

**Comment by matthiaskoenig** [comment]
Sorry attached the wrong model version. Here the one with uncertainties
[simvastatin_liver_uncertainty.zip](https://github.com/sys-bio/roadrunner/files/5473800/simvastatin_liver_uncertainty.zip)

**Comment by luciansmith** [comment]
I have seen this too, as a bug in llvm, though I found it with actually-used distributions, not ones from annotations.  

Try using https://github.com/sys-bio/llvm-6.x to see if it goes away (I removed the assert in question from our version of the source code), or try compiling llvm in release mode, which will also ignore the asserts.

If it still fails, let me know!

**Comment by luciansmith** [comment]
OK, more information:

* The various distribution functions are created for llvm if roadrunner finds the 'distrib' package being used at all.  I don't think this is worth changing, as the most common use-case will be using the distrib package to implement distrib functions.
* The problem is indeed an assert in llvm.  Unfortunately, I can't figure out what it is that causes it, but I did discover that ignoring the assert seems to be fine.  On Windows, compiling llvm in release mode works for this purpose, which makes me wonder if you're working in a system where you've compiled llvm yourself in debug mode?  If so, you can check out https://github.com/sys-bio/llvm-6.x and use that, because the assert in question has been removed.  If not, let us know what system you're using and we'll try to find a workaround.
* I saw that your model is not validatable in libsbml, so I implemented comp/distrib flattening so it would be.  https://github.com/sbmlteam/libsbml/pull/65

We have run into problems in compiling llvm on linux and mac, since it seems that the most recent compilers revealed a bug in llvm that I don't understand well enough to fix, that breaks testing.  Hopefully, we'll be able to find a workaround for it.

**Comment by matthiaskoenig** [comment]
Hi Lucian,
Thanks so much for the work. The distrib flattening is something I definitely need. In my workflow I checked that the model is valid without distrib and added the distrib information only afterwards.

I am using the roadrunner python package on Linux installed with pip (Ubuntu 20.04)
```
Linux trip3 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
[GCC 9.3.0] on linux
libroadrunner                 2.0.0
```
The LLVM should the LLVM which was used to compile the package. As I understand it these packages are build by your infrastructure so it should be the LLVM which was linked in the pip package for linux.
 
I could try compiling roadrunner myself with python packages, but this would only solve this issue on my machine (and have to figure out again how to do this ;) Is there an updated tutorial for building roadrunner?). I have students working with the same code base and they won't be able to compile roadrunner (these are biology students with minimal tech background).
So having a working pip package would be the best option. For testing I could compile this, but I can't use this in my workflows if my student can't run this.

As a bit of background information: I just started putting uncertainties and distributions into models and wanted to test these with roadrunner (basically as ground work for the distrib manuscript we shortly discussed during COMBINE). 

Best Matthias
